// gatsby/dist/isomorphic/HTML.js
'use strict';

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _typography = require('typography');

var _typography2 = _interopRequireDefault(_typography);

var TypographyStyle = new _typography2['default']().TypographyStyle;

module.exports = _react2['default'].createClass({
  displayName: 'exports',

  propTypes: {
    title: _react.PropTypes.string,
    body: _react.PropTypes.string,
    favicon: _react.PropTypes.string
  },

  getDefaultProps: function getDefaultProps() {
    return {
      title: 'Default title',
      body: ''
    };
  },

  render: function render() {
    return _react2['default'].createElement(
      'html',
      { lang: 'en' },
      _react2['default'].createElement(
        'head',
        null,
        _react2['default'].createElement('meta', { charSet: 'utf-8' }),
        _react2['default'].createElement('meta', { httpEquiv: 'X-UA-Compatible', content: 'IE=edge' }),
        _react2['default'].createElement('meta', { name: 'viewport', content: 'user-scalable=no width=device-width, initial-scale=1.0 maximum-scale=1.0' }),
        _react2['default'].createElement(
          'title',
          null,
          this.props.title
        ),
        _react2['default'].createElement('link', { rel: 'shortcut icon', href: this.props.favicon }),
        _react2['default'].createElement(TypographyStyle, null),
        _react2['default'].createElement('style', { dangerouslySetInnerHTML: { __html: 'body {color: rgb(66,66,66)} h1,h2,h3,h4,h5,h6 {color: rgb(44,44,44)}a {color: rgb(42,93,173)text-decoration: none}a:hover{text-decoration: underline}' } })
      ),
      _react2['default'].createElement(
        'body',
        { className: 'landing-page' },
        _react2['default'].createElement('div', { id: 'react-mount', dangerouslySetInnerHTML: { __html: this.props.body } }),
        _react2['default'].createElement('script', { src: '/bundle.js' })
      )
    );
  }
});
